### ISSUE TITLE
[R-package] cutomized loss function for survival data

### ISSUE THREAD

**Issue by yuebanfengqing** [dmlc/xgboost#5290]
I am trying to used xgboost package on the right-censored time-to-event data. It seems like there is no built-in loss function for survival data in R xgboost package. I tried to write mu own customized objective function, which is negative log partial likelihood. Since it has censoring indicator in it (besides, I have a treatment indicator), is it possible to pass those information into the customized loss function as an argument in defining DMatrix or the objective function?

Below is my code for the customized objective function:

  survobj <- function(preds, dtrain) {
  
    labels <- getinfo(dtrain, "label")
  
    n <- length(preds)
    r <- rank(labels)
  
    ita <- preds
    expita <- exp(ita)
  
    dono <- rep(0, n)
    grad.dono <- rep(0, n)
  
    for (i in 1:n) {
      dono[i] <- sum(expita[r >= r[i]])
      # derivative of dono
      grad.dono[i] <- sum(t.mod[r >= r[i]] * expita[r >= r[i]])
    }
  
    # gradient and hessian of the log partial likelihood
  
    # llk <- -(ita - log(dono))*event ### negative log partial likelihood
    grad <- -(t.mod - grad.dono / dono)*event / weights
    hess <- (dono^2 - grad.dono^2)/ dono^2 * event / weights
  
    return(list(grad = grad, hess = hess))
  }

Thanks a lot!

**Comment by yuebanfengqing** [comment]
If not, is there any ways to modify the code of xgb.DMatrix so that it can take more parameters?

**Comment by trivialfis** [comment]
@hcho3

**Comment by hcho3** [comment]
We added support for AFT survival regression in 1.1.0. See the tutorial at https://xgboost.readthedocs.io/en/latest/tutorials/aft_survival_analysis.html. In your customized objective, you should be able to access the lower and upper bounds on the label with
```
label_upper_bound <- getinfo(dtrain, "label_upper_bound")
label_lower_bound <- getinfo(dtrain, "label_lower_bound")
```
